### PR TITLE
ENT-2848 More contract attachment caching to avoid database queries slowing the node down

### PR DESCRIPTION
**Pull request by rick-r3**

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 1]
```diff
@@ -403,32 +402,62 @@ class NodeAttachmentService(
...
-                val criteriaBuilder = session.criteriaBuilder
+        fun toList(): List<AttachmentId> =
+            if(signed != null) {
+                if(unsigned != null) {
+                    listOf(signed, unsigned)
+                } else listOf(signed)
+            } else listOf(unsigned!!)
+    }
+
+    private val contractsCache = InfrequentlyMutatedCache<String, NavigableMap<Int, AttachmentIds>>("NodeAttachmentService_contractAttachmentVersions", cacheFactory)
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: Can you just add a comment here to explain the contents of this cache:
ContractClassName -> Versions -> AttachmentIds (where a single version of a ContractClassName may have one or both of a signed/unsigned version of the same contract jar code).

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: I've added a comment and some type aliases to make it clearer.

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 2]
```diff
@@ -45,6 +45,9 @@ import java.util.jar.JarInputStream
 import javax.annotation.concurrent.ThreadSafe
 import javax.persistence.*
 
+typealias ClassName = String
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: actually we already have both of these typealias's ;-)

`typealias ContractClassName = String` in `TransactionState` (core)
`typealias Version = Int` in `Version.kt` (core)

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: So we do!  Switched to those.

**Pull request merged**